---
title: "10_fig1-HYDG-tree"
title-block-banner: true
author:
  - name: "Jinlong Ru"
    orcid: "0000-0002-6757-6018"
    email: "jinlong.ru@gmail.com"
    affiliations:
      - name: "Institute of Virology, Helmholtz Centre Munich â€“ German Research Center for Environmental Health, 85764 Neuherberg, Germany"
      - neme: "Chair of Prevention of Microbial Diseases, School of Life Sciences Weihenstephan, Technical University of Munich, 85354 Freising, Germany"
date: "2022-11-06"
toc: true
format: gfm
bibliography: refs.bib
params:
  name: "10_fig1-HYDG-tree" # change if you rename file
---

Updated at `r format(Sys.time(), '%Y-%m-%d %H:%M:%S', tz = 'UTC')` UTC.

```{r setup, message=FALSE, include=F}
here::i_am(paste0(params$name, ".qmd"), uuid = "cdc28f4c-4364-4c58-8fa2-c91fc4d1dfd4")
projthis::proj_create_dir_target(params$name, clean = FALSE)
path_target <- projthis::proj_path_target(params$name)
path_source <- projthis::proj_path_source(params$name)
```

# Introduction

We have three phylogenetic trees for three protein families,

1. fam0023_tree.nwk
2. fam0026_tree.nwk
3. fam0253_tree.nwk

There is a table contains annotation of all the proteins in these trees. Annotations include taxonomy, environment etc. Expected output of this task is three phylogenetic trees, with taxonomy and environment annotations.

Trees can be plotted using `ggtree`. Annotations can be added using `ggtreeExtra`. Tutorial can be found [here](https://yulab-smu.top/treedata-book/chapter10.html). One example is [this](https://www.nature.com/articles/s41586-020-2007-4/figures/2).

## Tasks

TODO:

- [ ] Change layout of the tree if necessary

```{r}
library(here)
library(conflicted)
library(tidyverse)
library(treedataverse)
library(data.table)
library(tidytree)
library(ape)
library(ggtree)

dir_prj <- normalizePath("..")
wd <- here(dir_prj, "data/fig1_HYDG_tree")

# function to read the ".nwk" tree file, plot using ggtree, and save to pdf file
create_tree <- function(ftree, df_anno_imgvr, df_anno_nr) {
  tree_tbl <- read.tree(ftree) %>%
    as_tibble() %>%
    mutate(UVIG = ifelse(str_detect(label, "^IMGVR"), label, "")) %>%
    mutate(UVIG = str_replace_all(UVIG, "\\|.*", ""))

  tree_IMGVR <- tree_tbl %>% 
    dplyr::filter(UVIG != "") %>%
    left_join(df_anno_imgvr[, c("UVIG", "Kingdom", "Phylum", "Class", "eco3")], by = "UVIG") %>%
    mutate(Kingdom = "Viruses")

  tree_nr <- tree_tbl %>%
    dplyr::filter(UVIG == "") %>%
    left_join(df_anno_nr[, c("protid", "Kingdom", "Phylum", "Class")], by = c("label" = "protid")) %>%
    mutate(eco3 = "Unknown")

  tree <- rbind(tree_IMGVR, tree_nr) %>%
    as.treedata()

  p <- ggtree(tree, layout="equal_angle", color="grey") +
    geom_tiplab2(aes(label=Kingdom, color = Kingdom), size=0, color="black", offset=0) +
    geom_tippoint(aes(color=Kingdom), size=0.2, alpha=0.7) 
  # output file name is basename of ftree without extension
  ggsave(filename=path_target(paste0(basename(ftree), ".pdf")), plot=p, width = 20, height = 13)
}
```

### Protein annotation

Proteins from three sources,

1. IMGVR, which include query viral proteins from IMGVR database.
2. Custom data, which include query viral proteins form other sources.
3. NR, which include proteins of bacteria, archaea, eukaryota and viruses in NCBI NR database that is homologous of viral proteins in 1 and 2.

Annotations of NR are stored in `df_anno_nr`, and saved to file `annotation_nr.tsv`.

```{r protein host taxonomy annotation from NR}
# taxid2lineage <- fread(here(wd, "nf_tmp_taxaid2lineage.tsv.gz"), header=F) %>%
#   setnames(colnames(.), c("taxid", "Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"))
# 
# name2taxid <- fread(here(wd, "nf_tmp_name2taxid.tsv.gz"), header=F) %>%
#   setnames(colnames(.), c("taxa", "taxid")) %>%
#   # sort by taxid in ascending order
#   arrange(taxid) %>%
#   # drop duplicates by taxa
#   distinct(taxa, .keep_all = TRUE)
# 
# df_anno_nr <- fread(here(wd, "targets_taxa.tsv"), header = F) %>% 
#   setnames(colnames(.), c("protid", "taxa")) %>%
#   left_join(name2taxid, by = "taxa") %>%
#   left_join(taxid2lineage, by = "taxid") %>%
#   distinct(protid, .keep_all = TRUE) %>%
#   # replace NA with "Unknown" in Kingdom, Phylum, Class, Order, Family, Genus, Species
#   mutate(Kingdom = ifelse(is.na(Kingdom), "Unknown", Kingdom),
#          Phylum = ifelse(is.na(Phylum), "Unknown", Phylum),
#          Class = ifelse(is.na(Class), "Unknown", Class),
#          Order = ifelse(is.na(Order), "Unknown", Order),
#          Family = ifelse(is.na(Family), "Unknown", Family),
#          Genus = ifelse(is.na(Genus), "Unknown", Genus),
#          Species = ifelse(is.na(Species), "Unknown", Species))
# 
# # write protein taxonomy annotation to file
# fwrite(df_anno_nr, path_target("annotation_nr.tsv"), sep = "\t", quote = F, row.names = F)
df_anno_nr <- fread(here(wd, "annotation_nr.tsv.gz"))
```

Annotations of IMGVR and others are stored in `df_anno_imgvr`, and saved to file `annotation_imgvr.tsv`.

```{r tree node annotation of protines on 3096 IMGVR contigs}
# df_anno_imgvr <- fread(here(wd, "HYDGfam_contig_info2.tsv")) %>%
#   setnames(colnames(.), make.names(colnames(.))) %>%
#   mutate(Host.taxonomy.prediction = ifelse(Host.prediction.method=="Isolate taxonomy", Host.taxonomy.prediction, "")) %>%
#   dplyr::select(c(UVIG, Ecosystem.classification, Topology, Taxonomic.classification, Host.taxonomy.prediction)) %>%
#   # split Taxonomic.classification to Domain, Kingdom, Phylum, Class, Order, Family, Genus based on ";"
#   tidyr::separate(Taxonomic.classification, c("Domain", "Kingdom", "Phylum", "Class", "Order", "Family", "Genus"), sep = ";") %>%
#   # split Host.taxonomytidyr::.prediction to host_Kingdom, host_Phylum, host_Class, host_Order, host_Family, host_Genus, host_Species based on ";"
#   tidyr::separate(Host.taxonomy.prediction, c("host_Kingdom", "host_Phylum", "host_Class", "host_Order", "host_Family", "host_Genus", "host_Species"), sep = ";") %>%
#   # split Ecosystem.classification to eco1, eco2, eco3, eco4 based on ";"
#   tidyr::separate(Ecosystem.classification, c("eco1", "eco2", "eco3", "eco4"), sep = ";") %>%
#   # replace all "^.*__" with ""
#   mutate_all(funs(str_replace_all(., "^.*__", "")))
# 
# # write protein taxonomy annotation to file
# fwrite(df_anno_imgvr, path_target("annotation_imgvr.tsv"), sep = "\t", quote = F, row.names = F)
df_anno_imgvr <- fread(here(wd, "annotation_imgvr.tsv.gz"))
```


### Tree visualization

Phylogenetic tree of each protein family was inferred using the maximum likelihood method with the fasttree program [@2010-FastTree_Price]. The best-scoring ML tree was selected from 100 bootstrap replicates. Taxonomy annotation of each protein was obtained from the hit results and normalized using Taxonkit software [@2021-TaxonKit_Shen]. The tree was visualized using ggtree software [@2022-Ggtree_Xu].

TODO:

- [ ] Plot tree for fam0253

- [ ] Collapse tree to phylum level. That is, if nodes in the tree have same phylum, collapse them into one node. Because each tree contains thousands of nodes, we hope using this approach, we can have fewer nodes and make the tree easier to visualize.

- [ ] Add Phylum name, environment source as a circular heatmap in the outside of the tree.

```{r using alkb.nwk as test}
# create_tree(here(wd, "top_alkb_tree.nwk"), df_anno_imgvr, df_anno_nr)
```

Files were saved to `r path_target()`.

TODO:

- [ ] If the plot looks good, plot for all three trees

```{r tree visualization}
ftrees <- list.files(here(wd), pattern = "fam.*_tree.nwk", full.names = T) %>%
  # set names using famid
  setNames(stringr::str_extract(., "fam\\d+"))

# create tree visualization for each protein family using function `create_tree`
# very time-consuming
# lapply(ftrees, create_tree, df_anno_nr = df_anno_nr, df_anno_imgvr = df_anno_imgvr)
```


## Files written

These files have been written to the target directory, ```r paste0("data/", params$name)```:

```{r list-files-target}
projthis::proj_dir_info(path_target())
```

### References
